jina/drivers/__init__.py

```diff
@@ -216,20 +216,24 @@ def __getstate__(self) -> Dict[str, Any]:
 class BaseRecursiveDriver(BaseDriver):
 
     def __init__(self,
+                 recur_on: Union[str, Tuple[str, str]] = ('chunks', 'matches'),
```

Comment:
I'd prefer to rename chunks to documents because the apply() function take effects on the top level documents as well. When granularity=0, the documents are **NOT** chunks.

Reply:
But it is still run in the chunk direction. And matches are documents as well. From a data structure chunks and matches are equal (just different name). Thus having the same behavior in the recursion for both makes the idea of "they are orthogonal" even more true. And the documents at depth=0 and adjacency=0 are iterable via the chunk or the matches direction. This also explains the behavior you mentioned above: When `granularity_range = (0,0)` and `adjacency_range = (0,0)` the following think are equivalent: `recur_on=["matches"]`, `recur_on=["chunks"]` and `recur_on=["chunks", "matches]`. I made the last case explicitly equivalent, since a double `_apply_all` for a executor like `Reverse` would mean they are twice reversed and thus the same again. Also it somehow makes sense, that `apply` and `apply_all` is only called once for each document.